hivemq-edge/src/main/java/com/hivemq/edge/modules/adapters/impl/polling/PollingTask.java

```diff
@@ -0,0 +1,210 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class PollingTask implements Runnable {
+
+    private static final Logger log = LoggerFactory.getLogger(PollingTask.class);
+
+    private final @NotNull ProtocolAdapterPollingSampler sampler;
+    private final @NotNull ScheduledExecutorService scheduledExecutorService;
+    private final @NotNull EventService eventService;
+    private final @NotNull AtomicInteger watchdogErrorCount = new AtomicInteger();
+    private final @NotNull AtomicInteger applicationErrorCount = new AtomicInteger();
+
+    private volatile long nanosOfLastPolling;
+    private final @NotNull AtomicBoolean continueScheduling = new AtomicBoolean(true);
+
+
+    public PollingTask(
+            final @NotNull ProtocolAdapterPollingSampler sampler,
+            final @NotNull ScheduledExecutorService scheduledExecutorService,
+            final @NotNull EventService eventService) {
+        this.sampler = sampler;
+        this.scheduledExecutorService = scheduledExecutorService;
+        this.eventService = eventService;
+    }
+
+    @Override
+    public void run() {
+        try {
+            nanosOfLastPolling = System.nanoTime();
+
+            // synchronized with stop so that the change of the executor future field and the setting of the continueScheduling
+            // has exclusive access by either of the two methods.
+            synchronized (this) {
+                if (!continueScheduling.get()) {
+                    return;
+                }
+                final CompletableFuture<?> localExecutionFuture = sampler.execute()
+                        .orTimeout(InternalConfigurations.ADAPTER_RUNTIME_JOB_EXECUTION_TIMEOUT_MILLIS.get(),
+                                TimeUnit.MILLISECONDS);
+                localExecutionFuture.whenComplete((aVoid, throwable) -> {
+                    if (throwable == null) {
+                        reschedule(0);
+                    } else {
+                        if (ExceptionUtils.isInterruptedException(throwable)) {
+                            handleInterruptionException(throwable);
+                        } else {
+                            handleExceptionDuringPolling(throwable);
+                        }
+                    }
+                });
+            }
+        } catch (Throwable t) {
+            // the sampler shouldnt throw a exception, but better safe than sorry as we might to
+            handleExceptionDuringPolling(t);
+        }
+    }
+
+    public synchronized void stopScheduling() {
+        continueScheduling.set(false);
```

Comment:
Not sure if possible, but it would be nice to speed up a shutdown if we could interrupt the thread here, and not reschedule it. But I am not sure if we get access to the thread.

Reply:
that is the problem. Before we interrupted the threads of the service that were waiting on the future. However the thread that really did the work, was never touched as this thread could be any thread. It could be a thread started by the adapter, a fork-join-pool thread or anything basically. We do not need to interrupt as the tasks should be fast as we do not wait for the future anymore in the loop. Unless an adapter blocks in the poll() method, which no adapter should do, but use the async api we offer.